// ng-lightning/buttons/button-state.js
import { Directive, Input, Output, EventEmitter, HostListener, ElementRef, Renderer } from '@angular/core';
import { toBoolean } from '../util/util';
export var NglButtonState = (function () {
    function NglButtonState(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.selectedChange = new EventEmitter();
    }
    Object.defineProperty(NglButtonState.prototype, "selected", {
        get: function () {
            return this._selected;
        },
        set: function (_selected) {
            this._selected = toBoolean(_selected);
            this.toggleClass('slds-is-selected', this._selected);
            this.toggleClass('slds-not-selected', !this._selected);
        },
        enumerable: true,
        configurable: true
    });
    NglButtonState.prototype.onSelectChange = function () {
        this.selectedChange.emit(!this.selected);
    };
    NglButtonState.prototype.toggleClass = function (className, isAdd) {
        this.renderer.setElementClass(this.element.nativeElement, className, isAdd);
    };
    NglButtonState.decorators = [
        { type: Directive, args: [{
                    selector: '[nglButtonState]',
                    host: {
                        'aria-live': 'assertive',
                    },
                },] },
    ];
    /** @nocollapse */
    NglButtonState.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer, },
    ]; };
    NglButtonState.propDecorators = {
        'selected': [{ type: Input, args: ['nglButtonState',] },],
        'selectedChange': [{ type: Output, args: ['nglButtonStateChange',] },],
        'onSelectChange': [{ type: HostListener, args: ['click',] },],
    };
    return NglButtonState;
}());
;
