// ng-lightning/menus/dropdown-item.js
import { Directive, ElementRef, Renderer, HostListener } from '@angular/core';
export var NglDropdownItem = (function () {
    function NglDropdownItem(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.isFocused = false;
    }
    NglDropdownItem.prototype.onFocus = function () {
        this.isFocused = true;
    };
    NglDropdownItem.prototype.onBlur = function () {
        this.isFocused = false;
    };
    NglDropdownItem.prototype.hasFocus = function () {
        return this.isFocused;
    };
    NglDropdownItem.prototype.focus = function () {
        this.renderer.invokeElementMethod(this.element.nativeElement, 'focus', []);
    };
    NglDropdownItem.decorators = [
        { type: Directive, args: [{
                    selector: '[nglDropdownItem]',
                    host: {
                        'tabindex': '0',
                    },
                },] },
    ];
    /** @nocollapse */
    NglDropdownItem.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer, },
    ]; };
    NglDropdownItem.propDecorators = {
        'onFocus': [{ type: HostListener, args: ['focus',] },],
        'onBlur': [{ type: HostListener, args: ['blur',] },],
    };
    return NglDropdownItem;
}());
