// ng-lightning/images/avatar.js
import { Component, Input, ChangeDetectionStrategy, ElementRef, Renderer } from '@angular/core';
import { replaceClass } from '../util/util';
export var NglAvatar = (function () {
    function NglAvatar(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.src = '';
        this.alt = '';
        renderer.setElementClass(element.nativeElement, 'slds-avatar', true);
    }
    Object.defineProperty(NglAvatar.prototype, "setSize", {
        set: function (value) {
            this.updateClass(this._size, value);
            this._size = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NglAvatar.prototype, "setType", {
        set: function (value) {
            this.updateClass(this._type, value);
            this._type = value;
        },
        enumerable: true,
        configurable: true
    });
    NglAvatar.prototype.ngOnInit = function () {
        if (!this._type) {
            this.renderer.setElementClass(this.element.nativeElement, 'slds-avatar--rectangle', true);
        }
        if (!this._size) {
            this.renderer.setElementClass(this.element.nativeElement, 'slds-avatar--medium', true);
        }
    };
    NglAvatar.prototype.updateClass = function (oldValue, newValue) {
        replaceClass(this, "slds-avatar--" + oldValue, newValue ? "slds-avatar--" + newValue : '');
    };
    NglAvatar.decorators = [
        { type: Component, args: [{
                    selector: 'ngl-avatar',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: "\n    <img [src]=\"src\" [alt]=\"alt\">\n  ",
                },] },
    ];
    /** @nocollapse */
    NglAvatar.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer, },
    ]; };
    NglAvatar.propDecorators = {
        'src': [{ type: Input },],
        'alt': [{ type: Input },],
        'setSize': [{ type: Input, args: ['size',] },],
        'setType': [{ type: Input, args: ['type',] },],
    };
    return NglAvatar;
}());
;
