// ng-lightning/tabs/tab-verbose.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
import { Directive, Input, TemplateRef, Output, EventEmitter, ContentChild } from '@angular/core';
import { NglTab } from './tab';
/*
 * <ngl-tab [heading="..."]>
 *    <template ngl-tab-heading>...</template>
 *    <template ngl-tab-content>
 *       Content goes here...
 *    </template>
 * </ngl-tab>
 */
export var NglTabHeading = (function () {
    function NglTabHeading(templateRef) {
        this.templateRef = templateRef;
    }
    NglTabHeading.decorators = [
        { type: Directive, args: [{ selector: 'template[ngl-tab-heading]' },] },
    ];
    /** @nocollapse */
    NglTabHeading.ctorParameters = function () { return [
        { type: TemplateRef, },
    ]; };
    return NglTabHeading;
}());
export var NglTabContent = (function () {
    function NglTabContent(templateRef) {
        this.templateRef = templateRef;
    }
    NglTabContent.decorators = [
        { type: Directive, args: [{ selector: 'template[ngl-tab-content]' },] },
    ];
    /** @nocollapse */
    NglTabContent.ctorParameters = function () { return [
        { type: TemplateRef, },
    ]; };
    return NglTabContent;
}());
export var NglTabVerbose = (function (_super) {
    __extends(NglTabVerbose, _super);
    function NglTabVerbose() {
        _super.apply(this, arguments);
        this.onActivate = new EventEmitter();
        this.onDeactivate = new EventEmitter();
    }
    NglTabVerbose.prototype.ngAfterContentInit = function () {
        if (this.headingTemplate) {
            this.heading = this.headingTemplate.templateRef;
        }
        this.templateRef = this.contentTemplate.templateRef;
    };
    NglTabVerbose.decorators = [
        { type: Directive, args: [{
                    selector: 'ngl-tab',
                    providers: [{ provide: NglTab, useExisting: NglTabVerbose }],
                },] },
    ];
    /** @nocollapse */
    NglTabVerbose.ctorParameters = function () { return []; };
    NglTabVerbose.propDecorators = {
        'id': [{ type: Input, args: ['nglTabId',] },],
        'heading': [{ type: Input },],
        'onActivate': [{ type: Output },],
        'onDeactivate': [{ type: Output },],
        'contentTemplate': [{ type: ContentChild, args: [NglTabContent,] },],
        'headingTemplate': [{ type: ContentChild, args: [NglTabHeading,] },],
    };
    return NglTabVerbose;
}(NglTab));
